### ISSUE TITLE
Uncaught TypeError: Cannot read property 'x' of undefined when restoring elements

### ISSUE THREAD

**Issue by Melnorme1984** [cytoscape/cytoscape.js#2436]
I'm getting the same error as this gentleman: https://github.com/cytoscape/cytoscape.js/issues/2402

But in a different situation. I remove some nodes from a graph, then initialize a layout such that those nodes are restored when the layout has stopped. Like so:

```
var neighbors = graph.remove("node[type='neighbor']");

graph.layout({
    "name": "cola",
    "randomize": true,
    "stop": function () {
        neighbors.restore();
    }
}).run();
```

Something like this is supposed to work, no? Once a layout has stopped, nothing is supposed to be changing.

**Comment by Melnorme1984** [comment]
Wait, I'm actually not sure this has anything to do with the layout.

Is it possible that elements currently can't be restored at all?

**Comment by maxkfranz** [comment]
I ran this code on the [cola demo](http://js.cytoscape.org/demos/colajs-graph/) in the console:

```js
eles = cy.remove('#605755'); cy.layout({ name: 'cola', stop: () => eles.restore() }).run(); 
```

That removes the PCNA node and restores it in the same way as your code.  It doesn't seem to reproduce the issue.  I'd need a reproducible demo to look into this any further.

**Comment by Melnorme1984** [comment]
Sorry, I was busy with other things and was unable to attend to this.

Check out this JS Bin: https://jsbin.com/cedalavore/edit

If you update Cytoscape to version 3.4.3 or higher, it'll bug out.

**Comment by maxkfranz** [comment]
Nor would this minimal example, based on your demo, work:

```js
var graphElements = {"nodes":[{"data":{"name":"AT5G57460","id":"835851","symbol":"AT5G57460","type":"neighbor"}},{"data":{"name":"UBQ3","id":"831899","symbol":"UBQ3","type":"neighbor"}},{"data":{"name":"TAP46","id":"835379","symbol":"TAP46","type":"neighbor"}},{"data":{"name":"AT1G20760","id":"838666","symbol":"AT1G20760","type":"neighbor"}},{"data":{"name":"CYCA3;3","id":"841125","symbol":"CYCA3;3","type":"neighbor"}},{"data":{"name":"GRF1","id":"826479","symbol":"GRF1","type":"neighbor"}},{"data":{"name":"RPT2a","id":"829025","symbol":"RPT2a","type":"neighbor"}},{"data":{"name":"CYCH;1","id":"832824","symbol":"CYCH;1","type":"neighbor"}},{"data":{"name":"ATI2","id":"827659","symbol":"ATI2","type":"neighbor"}},{"data":{"name":"ZAT7","id":"823752","symbol":"ZAT7","type":"neighbor"}},{"data":{"name":"RbcX2","id":"832106","symbol":"RbcX2","type":"neighbor"}},{"data":{"name":"RbcX1","id":"825753","symbol":"RbcX1","type":"neighbor"}},{"data":{"name":"GRF3","id":"833836","symbol":"GRF3","type":"neighbor"}},{"data":{"name":"AT5G14240","id":"831274","symbol":"AT5G14240","type":"neighbor"}},{"data":{"name":"SCE1","id":"824956","symbol":"SCE1","type":"neighbor"}},{"data":{"name":"AT1G68185","id":"843147","symbol":"AT1G68185","type":"neighbor"}},{"data":{"name":"RAE1","id":"844406","symbol":"RAE1","type":"neighbor"}},{"data":{"name":"ABI5","id":"818199","symbol":"ABI5","type":"neighbor"}},{"data":{"name":"PIF4","id":"818903","symbol":"PIF4","type":"neighbor"}},{"data":{"name":"SWI3C","id":"838774","symbol":"SWI3C","type":"neighbor"}},{"data":{"name":"SWI3B","id":"817927","symbol":"SWI3B","type":"neighbor"}},{"data":{"name":"SWI3A","id":"819375","symbol":"SWI3A","type":"neighbor"}},{"data":{"name":"AT3G24150","id":"822001","symbol":"AT3G24150","type":"neighbor"}},{"data":{"name":"AT4G14720","id":"827125","symbol":"AT4G14720","type":"neighbor"}},{"data":{"name":"NTRC","id":"818766","symbol":"NTRC","type":"neighbor"}},{"data":{"name":"CPSF73-II","id":"814702","symbol":"CPSF73-II","type":"neighbor"}},{"data":{"name":"CLPS3","id":"819626","symbol":"CLPS3","type":"neighbor"}},{"data":{"name":"CPSF160","id":"835240","symbol":"CPSF160","type":"neighbor"}},{"data":{"name":"PCFS4","id":"825827","symbol":"PCFS4","type":"neighbor"}},{"data":{"name":"SUMO1","id":"828791","symbol":"SUMO1","type":"neighbor"}},{"data":{"name":"DDB1A","id":"825890","symbol":"DDB1A","type":"neighbor"}},{"data":{"name":"DDB1B","id":"827857","symbol":"DDB1B","type":"neighbor"}},{"data":{"name":"CPSF100","id":"832453","symbol":"CPSF100","type":"neighbor"}},{"data":{"name":"CPSF73-I","id":"842393","symbol":"CPSF73-I","type":"neighbor"}},{"data":{"name":"FY","id":"831191","symbol":"FY","type":"seed"}},{"data":{"name":"2-Cys Prx B","id":"830517","symbol":"2-Cys Prx B","type":"significant"}},{"data":{"name":"FCA","id":"827323","symbol":"FCA","type":"significant"}},{"data":{"name":"ATX","id":"841454","symbol":"THX","type":"seed"}},{"data":{"name":"TZ","id":"835567","symbol":"THI1","type":"seed"}},{"data":{"name":"AT3G11630","id":"820335","symbol":"AT3G11630","type":"significant"}}],"edges":[{"data":{"source":"835851","target":"820335","weight":0.16417210871}},{"data":{"source":"831899","target":"820335","weight":0.16417210871}},{"data":{"source":"835379","target":"820335","weight":0.16417210871}},{"data":{"source":"838666","target":"820335","weight":0.16417210871}},{"data":{"source":"841125","target":"820335","weight":0.16417210871}},{"data":{"source":"826479","target":"820335","weight":0.16417210871}},{"data":{"source":"829025","target":"820335","weight":0.16417210871}},{"data":{"source":"832824","target":"820335","weight":0.16417210871}},{"data":{"source":"827659","target":"820335","weight":0.426218332225}},{"data":{"source":"823752","target":"835567","weight":0.426218332225}},{"data":{"source":"832106","target":"835567","weight":0.16417210871}},{"data":{"source":"825753","target":"835567","weight":0.16417210871}},{"data":{"source":"833836","target":"835567","weight":0.16417210871}},{"data":{"source":"831274","target":"835567","weight":0.426218332225}},{"data":{"source":"824956","target":"841454","weight":0.426218332225}},{"data":{"source":"843147","target":"841454","weight":0.426218332225}},{"data":{"source":"844406","target":"841454","weight":0.16417210871}},{"data":{"source":"844406","target":"820335","weight":0.16417210871}},{"data":{"source":"818199","target":"827323","weight":0.306744265244}},{"data":{"source":"818903","target":"827323","weight":0.306744265244}},{"data":{"source":"838774","target":"827323","weight":0.426218332225}},{"data":{"source":"817927","target":"827323","weight":0.426218332225}},{"data":{"source":"819375","target":"827323","weight":0.426218332225}},{"data":{"source":"822001","target":"830517","weight":0.16417210871}},{"data":{"source":"822001","target":"835567","weight":0.16417210871}},{"data":{"source":"827125","target":"830517","weight":0.16417210871}},{"data":{"source":"827125","target":"835567","weight":0.16417210871}},{"data":{"source":"818766","target":"830517","weight":0.795566236042}},{"data":{"source":"818766","target":"841454","weight":0.306744265244}},{"data":{"source":"818766","target":"820335","weight":0.971612405345}},{"data":{"source":"814702","target":"831191","weight":0.16417210871}},{"data":{"source":"819626","target":"831191","weight":0.16417210871}},{"data":{"source":"835240","target":"831191","weight":0.63336666591}},{"data":{"source":"825827","target":"831191","weight":0.16417210871}},{"data":{"source":"828791","target":"831191","weight":0.16417210871}},{"data":{"source":"825890","target":"831191","weight":0.7436019184}},{"data":{"source":"827857","target":"831191","weight":0.16417210871}},{"data":{"source":"832453","target":"831191","weight":0.990058450158}},{"data":{"source":"842393","target":"831191","weight":0.16417210871}},{"data":{"source":"831191","target":"827323","weight":0.74962930614}},{"data":{"source":"830517","target":"841454","weight":0.795566236042}},{"data":{"source":"830517","target":"820335","weight":0.426218332225}},{"data":{"source":"841454","target":"820335","weight":0.795566236042}}]};

var graphStyle = [
    {
        "selector": "node[type='neighbor']",
        "style": {
            "display": "none"
        }
    }
];

var graph = window.cy = cytoscape({
    "container": document.getElementById('graph'),
    "elements": graphElements,
    "style": graphStyle,
    "layout": {"name": "grid"}
});
```

**Comment by maxkfranz** [comment]
It doesn't really make sense to include elements of undefined area in a layout.

**Comment by Melnorme1984** [comment]
So how would you suggest I implement this? I want these neighbor nodes to be invisible by default, and also for them not to be a factor in the initial cose layout (which is why I remove and then restore them only after it's finished calculating). If I understand correctly, what you're saying is that it doesn't make sense for a layout to handle invisible nodes, but in my case the layout has already finished its tasks, so why shouldn't that work?

The alternative here is what, calculating the neighbor nodes for each non-neighbor node separately, removing them all, and then making them appear by restoring them instead of setting their display attribute? That's a lot more complex...

**Comment by maxkfranz** [comment]
> what you're saying is that it doesn't make sense for a layout to handle invisible nodes

No, whether the element is visible is not the concern.  It's the area that the element takes up that matters.  Layouts are all about spacial area, so undefined element area does not make sense.  You can have a visibility:hidden element that takes up space but does not get shown.  Again, that's the same semantics as the DOM when a reflow (layout) is applied.

> The alternative here is what, calculating the neighbor nodes for each non-neighbor node separately, removing them all, and then making the appropriate set of neighbor nodes appear by restoring them instead of by setting their display attribute? That's rather more complex...

That's one alternative.  There are several others, such as applying a layout to only a subset of the graph or using different approaches to hide elements.  

See 

- http://js.cytoscape.org/#style/visibility
- http://js.cytoscape.org/#eles.layout

**Comment by Melnorme1984** [comment]
Ah yes, I guess the visibility attribute would do the trick.

Personally I think layouts should be robust enough to work with display: none'd elements and this should still count as a bug. Seems like it shouldn't be too hard to add some code to just ignore them, no? But I'm glad to have a workaround.

**Comment by Melnorme1984** [comment]
I haven't found that applying the layout to a specific subset of the graph changes anything. It seems the mere presence of display: none nodes in the graph is radioactive for layouts.

**Comment by maxkfranz** [comment]
The restrictions on fitting are similar to those of layout for this discussion.

I've reviewed this further and I've come to three conclusions:

(1) The main issue is a failure of the documentation.  It should be clearer what the semantics are for different ways of hiding elements.  It should also be clearer that you can run a layout on a subgraph, and common use-cases should be given for when that would be appropriate (e.g. excluding hidden elements).  Even if there were no exceptions in your demo (as in v3.4.2), the use of the API within that demo isn't right -- not yielding the intended behaviour.  More details are provided in #2471.

(2) There is indeed a problem with the library itself, which is demonstrated by your demo, introduced in commit 47360cae7ae8371d49e6d94824bd8f2f4379e563.  That commit does not use `ele.takesUpSpace()` in the bounds functions, which would be necessary to be consistent with the intention of that commit.  However, revising the library will not solve the root issue in your demo:  You're getting bad layout results, since you're running a layout on hidden elements.  Having said that, there are valid use-cases for running a layout on hidden elements.  Usually the intention there is to run a layout once on the entire graph so that showing a previously-hidden element requires no new layout:  The previously-hidden element is already in place.  More details are provided in #2472.

(3) The library already handles zero-area (e.g. display:none) elements w.r.t. fitting in a sensible way:  Zero-area elements are excluded from the composite bounding box.  The handling of zero-area elements -- especially for layout extensions that are third-party or that consume third-party libraries for the layout algorithm -- are not as robust as they should be.  More details are provided in #2473.

Because there are three separate concerns raised by your demo, I'm going to leave this issue closed.  Each concern can be addressed in a separate issue, cross-referenced with this issue.

**Comment by Melnorme1984** [comment]
It's been a while since I originally wrote this code, but I now recall that the reason I used display: none and not visibility: hidden is that it automatically hid the edges for me as well. So I've fixed this by switching to visibility and replacing calls to neighbors with neighbors.union(neighbors.connectedEdges()).

**Comment by maxkfranz** [comment]
Your `fit()` calls and `layout().run()` calls will also have to be on the visible subgraph rather than the whole graph.
